blog/2019-09-ngrx-data-views/README.md

```diff
@@ -0,0 +1,419 @@
+
+```ts
+book$ = this.store.pipe(select(getBook, this.bookId));
+```
+
+(The parameter `bookId` can come from the routing, be injected into the component or be a static reference. For this example it doesn't really matter. Have a look at the later following StackBlitz examples to see how we did it.)
+
+As the `getBook` selector only provides access to IDs of authors and tags, we have to compose additional selectors for the display of these entities. The one for the authors would look like this:
+
+```ts
+export const getAuthorsOfBook = createSelector(
+  getBook,
+  getAuthorEntities,
+  (book, authors): Author[] =>
+    book
+      ? book.authorIds.map(authorId => authors[authorId])
+      : []
+)
+```
+
+Have a look at how we used the previous selector `getBook` in the new selector. When using a parameterized selector in another selector, the new one also becomes a parameterized selector.
+
+The `getAuthorsOfBook` selector will be used in the component like this:
+
+```ts
+authors$ = this.store.pipe(select(getAuthorsOfBook, this.bookId));
+```
+
+Similarly a selector to retrieve the tags of the current book has to be composed.
+
+At the end we will have the following properties in our component file:
+
+```ts
+book$ = this.store.pipe(select(getBook, this.bookId));
+authors$ = this.store.pipe(select(getAuthorsOfBook, this.bookId));
+tags$ = this.store.pipe(select(getTagsOfBook, this.bookId));
+```
+
+And we will use them on the template like this:
+
+```html
+<div *ngIf="book$ | async as book">
+  <div>{{ book.title }}</div>
+  <div><span *ngFor="let a of (authors$ | async)">{{ a.lastName }}</span></div>
+  <div>{{ book.description }}</div>
+  <div><span *ngFor="let t of (tags$ | async)">{{ t.name }}</span></div>
+</div>
+```
+
+## The Problem
+
+You can probably see what this will lead to in the long run: Code duplication. Every component displaying books will have to select the normalized data from the store and accumulate it on the template. Now imagine an API change. With this amount of duplication it could mean that every components typescript file as well as all Angular templates would have to be touched. But how can we solve this?
+
+Abstractly speaking: Currently we retrieve the data normalized via the API (1) and put it normalized into the state (2). Then we use selectors (3) to retrieve the data normalized in each component and only then de-normalize it on each and every template (4) that displays it.
+
+![Normalization points](./normalization_points.svg)
+
+It seems there are three points where we could intervene and do the de-normalization instead:
+
+1. Retrieve the data ready for display:
+  
+   This can only be done by changing from a simple REST based API to an API that provides access to all data with one call. GraphQL would be an answer, but most often this API is not provided.
+
+2. Put the data de-normalized into the store:
+  
```

Comment:
Be more specific here about where the de-n11n takes place. "in the service"

Reply:
Hmm, I like to disagree. The fact that we at ISH use "services" only for something communicating to the API is quite limited. From a dependency injection point of view every instance in an injector is somewhat a "service". I really don't want to overuse this naming. :thinking:

Reply:
Okay, I meant something different 😅 This section just has one sentence and it's not clear at first glance what it means to "put the data de-normalized into the store". Maybe provide an example (and declare it as such): "This means that, if we have an HTTP service that fetches the data normalized, we use this service to build complex objects from fetched data, e.g. nested books and authors. Then we put this nested data into the store after fecthing it from HTTP"

Reply:
Ah, now I know what you mean. I'll take it ;)